src/vs/workbench/contrib/notebook/browser/notebookExecutionStateServiceImpl.ts

```diff
@@ -182,48 +187,31 @@ export class NotebookExecutionStateService extends Disposable implements INotebo
 		this._onDidChangeLastRunFailState.fire({ visible: false, notebook: notebookURI });
 	}
 
-	private _getFailedCellListener(notebookURI: URI) {
-		const notebook = this._notebookService.getNotebookTextModel(notebookURI);
-		let lastFailedCellDisposableListener;
-		if (notebook) {
-			lastFailedCellDisposableListener = combinedDisposable(
-				notebook.onWillAddRemoveCells((e: NotebookTextModelWillAddRemoveEvent) => {
-					const lastFailedCell = this._lastFailedCells.get(notebook.uri)?.cellHandle;
-					if (lastFailedCell !== undefined) {
-						const lastFailedCellPos = notebook.cells.findIndex(c => c.handle === lastFailedCell);
-						e.rawEvent.changes.some(([start, deleteCount]) => {
-							if (deleteCount) {
-								if (lastFailedCellPos >= start && lastFailedCellPos < start + deleteCount) {
-									this._setLastFailedCellVisibility(notebook.uri, false);
-									return true;
-								}
-							}
-							return false;
-						});
-					}
-				}),
-				notebook.onDidChangeContent((events) => {
-					const lastFailedCell = this._lastFailedCells.get(notebook.uri)?.cellHandle;
-					if (lastFailedCell !== undefined) {
-						events.rawEvents.forEach((e) => {
-							if (e.kind === NotebookCellsChangeType.ModelChange) {
-								e.changes.some(([start, deleteCount]) => {
-									if (deleteCount === 0) {
-										const addHandle = notebook.cells[start].handle;
-										if (addHandle === lastFailedCell) {
-											this._setLastFailedCellVisibility(notebook.uri, true);
-											return true;
-										}
-									}
-									return false;
-								});
+	private _getFailedCellListener(notebook: NotebookTextModel): IDisposable {
+		return this._register(
```

Comment:
Instead of calling register, it would be better to go through `this._lastFailedCells` in the dispose method, and dispose all the remaining cell infos there.

The difference is that I think it's best for a disposable to have a single clear "owner". If the disposable object has a lifecycle that is the same as this service, then call `this._register`. But these will usually be managed independently and disposed earlier, so it feels wrong to register them on the service.